controllers/post_controller.go

```diff
@@ -150,6 +150,142 @@ func GetMyPost(c *gin.Context) {
+	}
+	// Get the updated post data from the request body
+	var updatedPost models.Post
+	if err := c.ShouldBindJSON(&updatedPost); err != nil {
+		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+		return
+	}
+	// Update only the specified fields and retain the original values for the rest
+	if updatedPost.UserID != nil {
+		post.UserID = updatedPost.UserID
+	}
+	if updatedPost.Content != "" {
+		post.Content = updatedPost.Content
+	}
+	if !updatedPost.Date.IsZero() {
+		post.Date = updatedPost.Date
+	}
+	post.IsPublic = updatedPost.IsPublic
+	post.Category = updatedPost.Category
+	post.Date = currentDateTime
+	post.Like = updatedPost.Like
+	if len(post.Comment) == 0 {
+		post.Comment = []*firestore.DocumentRef{}
+	}
+	if len(post.Like) == 0 {
+		post.Like = []*firestore.DocumentRef{}
+	}
+
+	// Update the post document in the database
+	if _, err := client.Collection("Post").Doc(post_id).Set(c.Request.Context(), post); err != nil {
+		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
+		return
+	}
+
+	c.JSON(http.StatusOK, gin.H{"message": "Post updated successfully"})
+}
+
+// ---------------------------- hander function ----------------------------
+
+func deletePostAndRemoveReference(ctx context.Context, fsClient *firestore.Client, postID, userID string) error {
+	err := fsClient.RunTransaction(ctx, func(ctx context.Context, tx *firestore.Transaction) error {
+		// Get the post document
+		postDoc, err := tx.Get(fsClient.Doc("Post/" + postID))
+		if err != nil {
+			return err
+		}
+
+		// Get the user document
+		userDoc, err := tx.Get(fsClient.Doc("User/" + userID))
+		if err != nil {
+			return err
+		}
+
+		// Get the references from the user document as a map
+		referencesMap := make(map[string]interface{})
+		if err := userDoc.DataTo(&referencesMap); err != nil {
+			return err
+		}
+
+		references, ok := referencesMap["Posts"].([]interface{})
+		if !ok {
+			return fmt.Errorf("References field is not an array")
+		}
+
+		// Convert the references array to []*firestore.DocumentRef
+		refs := make([]*firestore.DocumentRef, len(references))
+		for i, ref := range references {
+			refs[i] = ref.(*firestore.DocumentRef)
+		}
+
+		// Find the index of the reference to the post
+		var index int
+		for i, ref := range refs {
+			if ref.ID == postID {
+				index = i
+				break
+			}
+		}
+
+		// Remove the reference from the references array
+		refs = append(refs[:index], refs[index+1:]...)
+
+		// Convert the references array back to an array of interface{}
+		newReferences := make([]interface{}, len(refs))
+		for i, ref := range refs {
+			newReferences[i] = ref
+		}
+
+		// Update the user document with the modified references array
+		referencesMap["Posts"] = newReferences
+		if err := tx.Set(userDoc.Ref, referencesMap, firestore.MergeAll); err != nil {
+			return err
+		}
+
+		// Delete the post document
+		if err := tx.Delete(postDoc.Ref); err != nil {
+			return err
+		}
+
+		return nil
+	})
+	return err
+}
```

Comment:
Do you have questions about this section? (that you refer)

Reply:
### step to delete a document in GOlang  ( It has 2 sections)

**Section 1 removes data in arrays Post (Collection User)**

1. find User information get data **section is Posts[]** ( arrays of posts in User Collection )
2. then get data to a local variable ( arrays name Posts[] )
3. and remove posts id within  arrays name Posts[] 
4. the last step in section 1, Use set into the same User Collection (that data new set  it will not post_id  appear )

**Section 2 removes document (Collection Post)**

1. normal delete by post_id and (like you do)

Reply:
``` suggestion
		// Convert the references array to []*firestore.DocumentRef
		refs := make([]*firestore.DocumentRef, len(references))
		for i, ref := range references {
			refs[i] = ref.(*firestore.DocumentRef)
		}
```
inline 253 - 255:  create local variable that's arrays

Reply:
line 259 - 266 : find index that you want to remove 
`
		// Find the index of the reference to the post
		var index int
		for i, ref := range refs {
			if ref.ID == postID {
				index = i
				break
			}
		}
`

Reply:
line 269 : create new array with remove id that you want 

```
		// Remove the reference from the references array
		refs = append(refs[:index], refs[index+1:]...)
```